### PR TITLE
fix sync issue that results in incorrect gradient accumulation and incorrect loss

### PR DESCRIPTION
**Pull request by WilsonCWu**
## Repro
Used torch version `'2.3.1+cu121'`.
Set `B = 16` (from 64).  
The following losses were observed:
```
250 val 6.4300
250 hella 0.2440
250 train 6.387966
...
1000 val 4.8797
1000 hella 0.2419
1000 train 4.924118
...
1250 val 4.5877
1250 hella 0.2526
1250 train 4.446614
...
2500 val 4.1250
2500 hella 0.2520
2500 train 3.954155
```

This should have been a purely mechanical change that doesn't affect loss (1 -> 4 grad accum steps), but comparing to @karpathy's run, the loss is clearly worse.

## Fix
For whatever reason, both the forward and backward pass need `require_backward_grad_sync=true` set in order to sync correctly. Found this thread with a similar observation: https://discuss.pytorch.org/t/whats-no-sync-exactly-do-in-ddp/170259

## Test
Running `B=16` with the patch shows the loss we expect.
```
250 val 6.1631
250 hella 0.2434
250 train 6.128294
...
1000 val 4.1533
1000 hella 0.2602
1000 train 4.170609
...
1250 val 3.9518
1250 hella 0.2559
1250 train 3.819895
```

This should fix training on 8x NVIDIA A100 SXM 40GB :) Note that you need B=16, 32 just barely doesn't fit.

### CODE REVIEW

**Comment by karpathy** [comment]
... TIL. I think nanoGPT is broken too, then.
